src/ophyd_async/epics/pvi/pvi.py

```diff
@@ -306,13 +316,25 @@ def create_children_from_annotations(
         if is_optional and name not in included_optional_fields:
             continue
         is_device_vector, device_type = _strip_device_vector(device_type)
+        if is_device_vector and not device_vectors:
+            continue
         if (
-            is_device_vector
+            (is_device_vector and name not in device_vectors.keys())
             or ((origin := get_origin(device_type)) and issubclass(origin, Signal))
             or (isclass(device_type) and issubclass(device_type, Signal))
         ):
             continue
 
-        sub_device = device_type()
-        setattr(device, name, sub_device)
-        create_children_from_annotations(sub_device)
+        if is_device_vector:
+            n_device_vector = DeviceVector(
+                {i: device_type() for i in range(1, device_vectors[name] + 1)}
+            )
+            setattr(device, name, n_device_vector)
+            for sub_device in list(n_device_vector.values()):
```

Comment:
Unnecessary list initialisation.

```suggestion
            for sub_device in n_device_vector.values():
```